---
title: Annotated reference data sets in _SingleR_
author: 
- name: Friederike DÃ¼ndar
  affiliation: Applied Bioinformatics Core, Weill Cornell Medicine
date: "Revised: August 14, 2019"
output:
  BiocStyle::html_document:
    toc_float: true
package: SingleR 
bibliography: ref.bib
vignette: >
  %\VignetteIndexEntry{2. Available reference datasets}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Introduction

The [legacy SingleR package](https://github.com/dviraran/SingleR/tree/master/data) provides RDA files that contain normalized expression values and cell types labels based on
bulk RNA-seq, microarray and single-cell RNA-seq data from:

* Blueprint and Encode (`blueprint_encode`),
* the Human Primary Cell Atlas (`hpca`),
* the murine [ImmGen](http://www.immgen.org/) (`immgen`), and
* a collection of mouse data sets downloaded from GEO (`mouse.rna.seq`).

The bulk RNA-seq and microarray data sets of the first three reference data sets were obtained from pre-sorted cell populations, i.e. the cell labels of these samples were mostly derived based on the respective sorting/purification strategy, not via *in silico* prediction methods.

| Data retrieval |	Organism	| Samples |	Sample types |	No. of main labels	| No. of fine labels |
|------------------|----------|----------|-------------|----------------------|------------|
|`HumanPrimaryCellAtlasData()`|	human	| 713	| microarrays of sorted cell populations	| 37 |	157 |
| `BlueprintEncodeData()` |	human	| 259	| RNA-seq |	24 | 43 |
|`ImmGenData()`|	mouse	| 830	 | microarrays of sorted cell populations|	20|	253|
|`MouseRNAseqData()`|	mouse	|358	|RNA-seq|	18	|28|

# Using a reference data set as training data with `SingleR`

If you want to use either one of these reference data sets as your training data 
set with the main `SingleR` function, you can locally cache the normalized expression
values and cell type labels with the respective functions shown here.
Each data set has a dedicated function: `HumanPrimaryCellAtlasData`, `BlueprintEncodeData`, `ImmGenData`, `MouseRNAseqData`.

The data retrieval functions will return a single R object of the `SummarizedExperiment` 
class with the normalized expression values stored in a gene-by-sample matrix in 
the `assay` slot (`"logcounts"`) and the sample labels as part of `colData(SE)`.
The general usage is shown here for data from HPCA:

```{r}
library(SingleR)
hpca.se <- HumanPrimaryCellAtlasData()

## example scRNA-seq data
library(scRNAseq)
hESCs <- LaMannoBrainData('human-es')
hESCs <- hESCs[,1:100] # for demo-purposes only!

## focus on genes that are common between your scRNA-seq and the reference data
library(scater)
common <- intersect(rownames(hESCs), rownames(hpca.se))
hpca.se <- hpca.se[common,]
hESCs <- hESCs[common,]
hESCs <- logNormCounts(hESCs)

## determine labels for the scRNA-seq data at hand using the reference data 
pred.hpca <- SingleR(test = hESCs, ref = hpca.se, labels = hpca.se$label.main)
table(pred.hpca$labels)
```

# Mouse reference data

## The Immunological Genome Project (ImmGen)

> The Immunological Genome Project is a collaborative group of immunologists and computational biologists who are generating a complete microarray dissection of gene expression and its regulation in the immune system of the mouse. The project encompasses the innate and adaptive immune systems, surveying all cell types of the myeloid and lymphoid lineages with a focus on primary cells directly *ex vivo*. [Source](http://www.immgen.org/)

830 microarray samples of ImmGen raw expression data (phases 1 and 2) were downloaded as CEL files from the Gene Expression Omnibus (GEO; GSE15907 and GSE37448), processed, and normalized 
using the robust multi-array average (RMA) procedure on probe-level data using Matlab functions.
The analysis was performed using a custom CDF file obtained from Brainarray  [@aran2019reference].

There are **20 main cell types** and **253 cell subtypes**.

| Label type | Labels |
|------------|--------|
| main | B cells, B cells, pro, Basophils, DC, Endothelial cells, Eosinophils, Epithelial cells, Fibroblasts, ILC, Macrophages, Mast cells, Microglia, Monocytes, Neutrophils, NK cells, NKT, Stem cells, Stromal cells, T cells, Tgd |
| fine | e.g. DC (DC.11B-), DC (DC.103-11B+24+), Macrophages (MF.103-11B+.SALM3, Macrophages (MF.11CLOSER.SALM3), Tgd (Tgd.vg2+.act)|

## Mouse RNA-seq samples of pure cell populations

The dataset entails **358 mouse RNA-seq** samples annotated to **18 main cell types** and
**28 fine-grained cell labels**.

| Label type | Labels |
|------------|--------|
| main | Adipocytes, Astrocytes, B cells, Cardiomyocytes, Dendritic cells, Endothelial cells, Epithelial cells, Erythrocytes, Fibroblasts, Granulocytes, Hepatocytes, Macrophages, Microglia, Monocytes, Neurons, NK cells, Oligodendrocytes, T cells |
| fine | Adipocytes, aNSCs, Astrocytes, Astrocytes activated, B cells, Cardiomyocytes, Dendritic cells, Endothelial cells, Ependymal, Erythrocytes, Fibroblasts, Fibroblasts activated, Fibroblasts senescent, Granulocytes, Hepatocytes, Macrophages, Macrophages activated, Microglia, Microglia activated, Monocytes, Neurons, Neurons activated, NK cells, NPCs, Oligodendrocytes, OPCs, qNSCs, T cells |

The RNA-seq data were downloaded from GEO and processed by the Benayoun Lab as follows [@Benayoun2019]:

>Paired-end 75-bp reads were trimmed using Trim Galore! 0.3.1 (github.com/FelixKrueger/TrimGalore) to retain high-quality bases with phred score >15 and a remaining length >35 bp. Read pairs were mapped to the UCSC mm9 genome build using STAR 2.4.0j (Dobin et al. 2013). Read counts were assigned to genes using subread 1.4.5-p1 (Liao et al. 2014) and were imported into R to estimate differential gene expression as a function of age using the DESeq2 R package (DESeq2 1.6.3). 

More details can be found in the [Benayoun Lab github repo](https://github.com/BenayounLaboratory/Mouse_Aging_Epigenomics_2018/tree/master/FigureS7_CIBERSORT).

# Human reference data

## Blueprint and ENCODE

Blueprint: 144 bulk RNA-seq of pure immune samples annotated to 28 cell types [@blueprintRef]

ENCODE: 115 RNA-seq pure stroma and immune samples annotated to 17 cell types [@encodeRef]

Altogether, 259 samples with **24 main labels** and  **43 "fine" cell labels**.

| Label type | Labels |
|------------|--------|
| main | Adipocytes, B-cells, CD4+ T-cells, CD8+ T-cells, Chondrocytes, DC, Endothelial cells, Eosinophils, Epithelial cells, Erythrocytes, Fibroblasts, HSC, Keratinocytes, Macrophages, Melanocytes, Mesangial cells, Monocytes, Myocytes, Neurons, Neutrophils, NK cells, Pericytes, Skeletal muscle, Smooth muscle |
| fine | Adipocytes, Astrocytes, CD4+ T-cells, CD4+ Tcm, CD4+ Tem, CD8+ T-cells, CD8+ Tcm, CD8+ Tem, Chondrocytes, Class-switched memory B-cells, CLP, CMP, DC, Endothelial cells, Eosinophils, Epithelial cells, Erythrocytes, Fibroblasts, GMP, HSC, Keratinocytes, Macrophages, Macrophages M1, Macrophages M2, Megakaryocytes, Melanocytes, Memory B-cells, MEP, Mesangial cells, Monocytes, MPP, mv Endothelial cells, Myocytes, naive B-cells, Neurons, Neutrophils, NK cells, Pericytes, Plasma cells, Preadipocytes, Skeletal muscle, Smooth muscle, Tregs |

## Human Primary Cell Atlas

Human Primary Cell Atlas (HPCA): a collection of Gene Expression Omnibus (GEO datasets), which
contains **713 microarray samples** classified to **37 main cell types** and further annotated to **157 subtypes** [@hpcaRef].

| Label type | Labels |
|------------|--------|
| main | Astrocyte, B_cell, BM, BM & Prog., Chondrocytes, CMP, DC, Embryonic_stem_cells, Endothelial_cells, Epithelial_cells, Erythroblast, Fibroblasts, Gametocytes, GMP, Hepatocytes, HSC_-G-CSF, HSC_CD34+, iPS_cells, Keratinocytes, Macrophage, MEP, Monocyte, MSC, Myelocyte, Neuroepithelial_cell, Neurons, Neutrophil, Neutrophils, NK_cell, Osteoblasts, Platelets, Pre-B_cell_CD34-, Pro-B_cell_CD34+, Pro-Myelocyte, Smooth_muscle_cells, T_cells, Tissue_stem_cells |
| fine | e.g. DC:monocyte-derived:AM580, DC:monocyte-derived:Poly(IC), T_cell:CD8+_naive, Keratinocytes:IL22, Keratinocytes:KGF |

# References
