---
title: SingleR-provided reference data sets
author: Friederike Duendar
date: "2019-08-08"
output:
  BiocStyle::html_document:
    toc_float: true
package: SingleR 
bibliography: ref.bib
vignette: >
  %\VignetteIndexEntry{Using SingleR to annotate scRNA-seq data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

The [legacy SingleR package](https://github.com/dviraran/SingleR/tree/master/data) provides the following RDA files:

`immgen`, `mouse.rna.seq`, `blueprint_encode`, `hpca`

These refer to Robjects with normalized expression values from large'ish collections
of human (`blueprint_encode`, `hpca`) and mouse (`immgen`, `mouse.rna.seq`) samples
and represent bulk RNA-seq (and microarray) data from purified cell populations.
That means that the cell labels of the samples were mostly derived based on the
respective sorting/purification strategy, not via *in silico* prediction methods.

If you want to use either one of these reference data sets as your training data 
set with the main `SingleR` function, you can locally cache the normalized expression
values and cell type labels with the respective functions shown here.
The data retrieval functions will return a single R object of the `SummarizedExperiment` 
class with the normalized expression values stored in a gene-by-sample matrix in 
the `assay` slot (`"normcounts"`) and the sample labels as part of `colData(SE)`.

Each reference collection comes with its own data retrieval function.
The general usage is shown here for data from HPCA:

```
library(SingleR)
## retrieve the reference data from the HPCA
hpca.se <- HumanPrimaryCellAtlasData()

## use the norm. expression values & cell type labels to assign cell type labels
## to the cells contained within <your_SCE>
predictLabels_with_hpca <- SingleR(test = your_SCE,
                             training = assay(hpca.se, "normcounts"), 
                             labels = colData(hpca.se)$label.main)
```


## Mouse reference data

### The Immunological Genome Project (ImmGen)

>The Immunological Genome Project is a collaborative group of immunologists and computational biologists who are generating a complete microarray dissection of gene expression and its regulation in the immune system of the mouse. The project encompasses the innate and adaptive immune systems, surveying all cell types of the myeloid and lymphoid lineages with a focus on primary cells directly *ex vivo*. [Source](http://www.immgen.org/)

830 microarray samples of ImmGen raw expression data (phases 1 and 2) were downloaded as CEL files from the Gene Expression Omnibus (GEO; GSE15907 and GSE37448), processed, and normalized 
using the robust multi-array average (RMA) procedure on probe-level data using Matlab functions.
The analysis was performed using a custom CDF file obtained from Brainarray  [@aran2019reference].

There are **20 main cell types** (B cells, B cells, pro, Basophils, DC, Endothelial cells, Eosinophils, Epithelial cells, Fibroblasts, ILC, Macrophages, Mast cells, Microglia, Monocytes, Neutrophils, NK cells, NKT, Stem cells, Stromal cells, T cells, Tgd) and **253 cell subtypes**.

```{r eval=FALSE}
library(SingleR)
## retrieve the reference data from the HPCA
immgen.se <- ImmGenData()

## see the main cell type labels
sort(unique(immgen.se$label.main))

## see the >253 immunological cell type labels
sort(unique(immgen.se$label.fine))

## use the norm. expression values & cell type labels of your choice to predict
## the cells present in your scRNA-seq data set
predictLabels_with_immgen <- SingleR(test = your_SCE,
                             training = assay(immgen.se, "normcounts"), 
                             labels = colData(immgen.se)$label.main)
```

### Benayoun et al.: mouse bulk RNA-seq 

This data set was contributed by the Benayoun Lab; it was generated as part of 
a [detailed study of aging mice](https://www.ncbi.nlm.nih.gov/pubmed/30858345) and 
the corresponding changes of their transcriptomes and epigenomes [@Benayoun2019].

The dataset entails **358 mouse RNA-seq** samples annotated to **18 main cell types**: Adipocytes, Astrocytes, B cells, Cardiomyocytes, Dendritic cells, Endothelial cells, Epithelial cells, Erythrocytes, Fibroblasts, Granulocytes, Hepatocytes, Macrophages, Microglia, Monocytes, Neurons, NK cells, Oligodendrocytes, T cells.

The more **28 fine-grained cell labels** encompass Adipocytes, aNSCs, Astrocytes, Astrocytes activated, B cells, Cardiomyocytes, Dendritic cells, Endothelial cells, Ependymal, Erythrocytes, Fibroblasts, Fibroblasts activated, Fibroblasts senescent, Granulocytes, Hepatocytes, Macrophages, Macrophages activated, Microglia, Microglia activated, Monocytes, Neurons, Neurons activated, NK cells, NPCs, Oligodendrocytes, OPCs, qNSCs, T cells.

The RNA-seq data were generated and processed as follows:

>We collected several tissues and cells from C57BL/6N male mice at three different time points: youth (3 mo), middle age (12 mo), and old age (29 mo). We focused on a subset of tissues—heart, liver, cerebellum, and olfactory bulb—that are known to display age-related functional decline and are clearly anatomically defined. We also derived primary cultures of neural stem cells (NSCs) from these young, middle-aged, and old mice. For each tissue or cell culture from all three ages, we generated transcriptomic maps (i.e., RNA-seq).

>Paired-end 75-bp reads were trimmed using Trim Galore! 0.3.1 (github.com/FelixKrueger/TrimGalore) to retain high-quality bases with phred score >15 and a remaining length >35 bp. Read pairs were mapped to the UCSC mm9 genome build using STAR 2.4.0j (Dobin et al. 2013). Read counts were assigned to genes using subread 1.4.5-p1 (Liao et al. 2014) and were imported into R to estimate differential gene expression as a function of age using the DESeq2 R package (DESeq2 1.6.3). [@Benayoun2019]

## Human reference data

### Blueprint and ENCODE

Blueprint: 144 RNA-seq pure immune samples annotated to 28 cell types [@blueprintRef]
ENCODE: 115 RNA-seq pure stroma and immune samples annotated to 17 cell types [@encodeRef]

Altogether, 259 samples classified to 43 cell types.

### Human Primary Cell Atlas

Human Primary Cell Atlas (HPCA): a collection of Gene Expression Omnibus (GEO datasets), which
contains 713 microarray samples classified to 38 main cell types and further annotated to 169 subtypes [@hpcaRef].

```{r eval=FALSE}
hpca.se <- HumanPrimaryCellAtlasData()
```


# References

