---
title: Using SingleR to annotate single-cell RNA-seq data
author: Aaron Lun
date: "Revised: 17 October 2018"
output:
  BiocStyle::html_document:
    toc_float: true
package: SingleR 
bibliography: ref.bib
vignette: >
  %\VignetteIndexEntry{Using SingleR to annotate scRNA-seq data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Introduction

`r Biocpkg("SingleR")` is an automatic annotation method for single-cell RNA sequencing (scRNAseq) data [@aran2019reference].
Given a training dataset of cells with known labels, it labels new cells from a test dataset based on similarity to the training set.
Specifically, for each test cell:

1. We compute the Spearman correlation between its expression profile and that of each training cell a given label.
2. We define the per-label score as a fixed quantile (by default, 0.8) of the distribution of correlations.
3. We repeat this for all labels and we take the label with the highest score as the annotation for this cell.
4. We optionally perform a fine-tuning step:
  - The training dataset is subsetted to only include labels with scores close to the maximum.
  - Scores are recomputed using only marker genes for the subset of labels.
  - This is iterated until one label remains.

Automatic annotation provides a convenient way of transferring biological knowledge across datasets.
In this manner, the burden of interpreting clusters and defining marker genes only has to be done once.

# Demonstration on pancreas data

## Obtaining the data 

Here, we will use two human pancreas datasets from the `r Biocpkg("scRNAseq")` package for training and testing [@muraro2016singlecell;@grun2016denovo].
We restrict our analysis to the common genes between the two datasets, also removing spike-in transcripts for simplicity.

```{r}
library(scRNAseq)
sceM <- MuraroPancreasData()
sceG <- GrunPancreasData()

common <- intersect(rownames(sceM), rownames(sceG))
common <- common[-grep("^ERCC-", common)]
sceM <- sceM[common,]
sceG <- sceG[common,]
```

One should normally do quality control at this point, but for brevity's sake, we will just remove the empty libraries here.
We then compute log-normalized expression values^[This is only necessary for the marker gene detection step.].

```{r}
library(scater)
sceM <- sceM[,!is.na(sceM$label)]
sceM <- normalize(sceM)
sceG <- sceG[,colSums(counts(sceG)) > 0]
sceG <- normalize(sceG)
```

The Muraro dataset contains labels so we will use this as our training dataset.

```{r}
table(sceM$label)
```

## Using default marker detection

The `SingleR()` function annotates a test dataset given a training dataset with known labels.
It returns a `DataFrame` containing the predicted labels and the per-label scores associated with each prediction.

```{r}
library(SingleR)
pred <- SingleR(test=sceG, training=sceM, labels=sceM$label, 
    assay.type="logcounts")
table(pred$labels)
```

By default, the `genes="de"` setting will identify markers based on pairwise comparisons between labels.
Specifically, `SingleR()` will compute the difference in the median log-expression between each pair of labels.
Markers are defined from the top set of genes with the largest positive differences in each pairwise comparison.
Correlations are then computed using only these informative genes.
Another strategy is to set `genes="sd"`, which will define markers based on the standard deviation of the median expression across labels.

## Using custom markers

Users can achieve more control of the annotation procedure by defining their own marker list.
This requires a series of pairwise comparisons between labels to define markers that distinguish each label from another.
For example, we can perform pairwise $t$-tests using `r Biocpkg("scran")` to obtain a list of lists of character vectors.
Each character vector contains the top 10 marker genes from a comparison between a pair of labels.

```{r}
library(scran)
out <- pairwiseTTests(logcounts(sceM), sceM$label, direction="up")

# To be moved into scran::topPairwiseMarkers().
markers <- list()
for (first in unique(sceM$label)) {
    cur.markers <- list()
    for (second in unique(sceM$label)) {
        chosen <- which(out$pairs$first==first & out$pairs$second==second)
        if (!length(chosen)) {
            cur.markers[[second]] <- character(0)
        } else {
            cur.stats <- out$statistics[[chosen]]
            cur.stats <- cur.stats[order(cur.stats$p.value),]
            cur.markers[[second]] <- head(rownames(cur.stats), 10)
        }
    }
    markers[[first]] <- cur.markers
}
```

We then supply these genes to `SingleR()` directly via the `genes=` argument.
This changes the results quite a bit, reflecting the importance of feature selection for this annotation method^[And many others.].

```{r}
pred2 <- SingleR(test=sceG, training=sceM, labels=sceM$label, genes=markers)
table(pred$labels, pred2$labels)
```

In some cases, markers may only be available for specific labels rather than for pairwise comparisons between labels.
This is accommodated by supplying a named list of character vectors to `genes`.
Note that this is likely to be less powerful than the list-of-lists approach as information about pairwise differences is discarded.

```{r}
label.markers <- lapply(markers, unlist, recursive=FALSE)
pred3 <- SingleR(test=sceG, training=sceM, labels=sceM$label, genes=label.markers)
table(pred$labels, pred3$labels)
```

# Available reference datasets

`r Biocpkg("SingleR")` can also directly download a number of training datasets generated from bulk RNA-seq of pure cell types.
This makes use of Immgen, Blueprint+Encode, the Human Primary Cell Atlas and other resources. 
Each reference dataset is obtained directly from the legacy SingleR Github repository by using the command:

```{r}
hpca <- getReferenceDataset(dataset="hpca")
```

This can then be directly used for annotation of our scRNA-seq dataset. 
Given the large number of samples (often from very disparate lineages) in each datset, providing the `main_types` list to the `labels=` argument can be helpful.
In this case, we will allow the marker genes to be computed and classify cell types for differentiated hESCs with the more specific `types` to the `labels=` argument.

```{r}
hESCs <- scRNAseq::LaMannoBrainData('human-es')

# Retain common genes and normalize as above.
common <- intersect(rownames(hESCs), rownames(hpca$data))
hESCs <- hESCs[common,]
hESCs <- hESCs[,colSums(counts(hESCs)) > 0]
hESCs <- scater::normalize(hESCs)
hpca$data <- hpca$data[common,]
pred.hpca <- SingleR(test=hESCs, training=hpca$data, labels=hpca$types)
```

# Separating training from classification

Advanced users can split the `SingleR()` workflow into two separate training and classification steps.
This means that training (e.g., marker detection, assembling of nearest-neighbor indices) only needs to be performed once.
The resulting data structures can then be re-used across multiple classifications with different test datasets, provided the test feature set is identical to or a superset of the features in the training set.
For example:

```{r}
trained <- trainSingleR(sceM, labels=sceM$label, genes=markers)
pred2b <- classifySingleR(sceG, trained)
table(pred2$labels, pred2b$labels)
```

Other efficiency improvements are possible through several arguments:

- Switching to an approximate algorithm for the nearest neighbor search in `trainSingleR()` via the `BNPARAM=` argument from the `r Biocpkg("BiocNeighbors")` package.
- Parallelizing the fine-tuning step in `classifySingleR()` with the `BPPARAM=` argument from the `r Biocpkg("BiocParallel")` package.

These arguments can also be specified in the `SingleR()` command.

# Session information

```{r}
sessionInfo()
```

# References
