---
title: Using SingleR to annotate single-cell RNA-seq data
author: 
- name: Aaron Lun 
- name: Jared M. Andrews
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
- name: Friederike DÃ¼ndar
  affiliation: Applied Bioinformatics Core, Weill Cornell Medicine
- name: Daniel Bunis
  affiliation: Bakar Computational Health Sciences Institute, University of California San Francisco, San Francisco, CA
date: "Revised: August 27, 2019"
output:
  BiocStyle::html_document:
    toc_float: true
package: SingleR 
bibliography: ref.bib
vignette: >
  %\VignetteIndexEntry{Annotating scRNA-seq data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
# library(BiocStyle)
```

# Introduction

`r Biocpkg("SingleR")` is an automatic annotation method for single-cell RNA sequencing (scRNAseq) data [@aran2019reference].
Given a reference dataset of samples (single-cell or bulk) with known labels, it labels new cells from a test dataset based on similarity to the reference set.
Specifically, for each test cell:

1. We compute the Spearman correlation between its expression profile and that of each reference sample. 
2. We define the per-label score as a fixed quantile (by default, 0.8) of the distribution of correlations.
3. We repeat this for all labels and we take the label with the highest score as the annotation for this cell.
4. We optionally perform a fine-tuning step:
  - The reference dataset is subsetted to only include labels with scores close to the maximum.
  - Scores are recomputed using only marker genes for the subset of labels.
  - This is iterated until one label remains.

Automatic annotation provides a convenient way of transferring biological knowledge across datasets.
In this manner, the burden of interpreting clusters and defining marker genes only has to be done once.

# Using the built-in references

`r Biocpkg("SingleR")` provides several reference datasets (mostly derived from bulk RNA-seq or microarray data) through dedicated data retrieval functions. 
For example, to obtain reference data from the Human Primary Cell Atlas:

```{r}
library(SingleR)
hpca.se <- HumanPrimaryCellAtlasData()
hpca.se
```

We use this reference in the `SingleR()` function to annotate a scRNA-seq dataset from @lamanno2016molecular.
For the sake of speed, we will only label the first 100 cells from this dataset.

```{r}
library(scRNAseq)
hESCs <- LaMannoBrainData('human-es')
hESCs <- hESCs[,1:100] 

# Restrict to common genes between test and reference data:
library(scater)
common <- intersect(rownames(hESCs), rownames(hpca.se))
hpca.se <- hpca.se[common,]
hESCs <- hESCs[common,]
hESCs <- logNormCounts(hESCs)

pred.hpca <- SingleR(test = hESCs, ref = hpca.se, labels = hpca.se$label.main)
pred.hpca
```

Each row of the output `DataFrame` contains prediction results for a single cell.
Labels are shown before fine-tuning (`first.labels`), after fine-tuning (`labels`) and after pruning (`pruned.labels`), along with the associated scores.
We summarize the distribution of labels across our subset of cells:

```{r}
table(pred.hpca$labels)
```

At this point, it is worth noting that `r Biocpkg("SingleR")` is workflow/package agnostic.
The above example uses `SummarizedExperiment` objects, but the same functions will accept any (log-)normalized expression matrix. 

# Using single-cell references

## Obtaining the data 

Here, we will use two human pancreas datasets from the `r Biocpkg("scRNAseq")` package [@muraro2016singlecell;@grun2016denovo].
The aim is to use one pre-labelled dataset to annotate the other unlabelled dataset.

```{r}
library(scRNAseq)
sceM <- MuraroPancreasData()
sceG <- GrunPancreasData()

# Restrict to common genes:
common <- intersect(rownames(sceM), rownames(sceG))
sceM <- sceM[common,]
sceG <- sceG[common,]
```

One should normally do cell-based quality control at this point, but for brevity's sake, 
we will just remove the empty or unlabelled libraries here.

```{r}
sceM <- sceM[,!is.na(sceM$label)]
sceG <- sceG[,colSums(counts(sceG)) > 0]
```

We then compute the log-normalized expression values that will be used by `SingleR()` for annotation.

```{r}
library(scater)
sceM <- logNormCounts(sceM)

# Some of the cells have no spike-in counts, so we 
# set use_altexps=FALSE to avoid normalization errors.
# This is fine as we don't use the spike-ins anyway.
sceG <- logNormCounts(sceG, use_altexps=FALSE)
```

The @muraro2016singlecell dataset contains labels so we will use this as our reference dataset.

```{r}
table(sceM$label)
```

To speed up this demonstration, we will subset the @grun2016denovo dataset to the first 100 cells.

```{r}
sceG <- sceG[,1:100]
```

## Using custom markers

The default marker definition in `SingleR()` is intended for references derived from bulk RNA-seq data.
When using single-cell data as a reference, we suggest building your own marker list.
This involves a series of pairwise comparisons between labels to define markers that distinguish each label from another,
and is easy to perform with functions from `r Biocpkg("scran")`.
For example, we can perform pairwise $t$-tests and obtain the top 10 marker genes from each pairwise comparison.

```{r}
library(scran)
out <- pairwiseTTests(logcounts(sceM), sceM$label, direction="up")
markers <- getTopMarkers(out$statistics, out$pairs, n=10)
```

We then supply these genes to `SingleR()` directly via the `genes=` argument.
A more focused gene set also allows annotation to be performed more quickly compared to the default approach.

```{r}
pred <- SingleR(test=sceG, ref=sceM, labels=sceM$label, genes=markers)
table(pred$labels)
```

In some cases, markers may only be available for specific labels rather than for pairwise comparisons between labels.
This is accommodated by supplying a named list of character vectors to `genes`.
Note that this is likely to be less powerful than the list-of-lists approach as information about pairwise differences is discarded.

```{r}
label.markers <- lapply(markers, unlist, recursive=FALSE)
pred2 <- SingleR(test=sceG, ref=sceM, labels=sceM$label, genes=label.markers)
table(pred$labels, pred2$labels)
```

# Annotation diagnostics

## Initial visualization

`r Biocpkg("SingleR")` provides a few basic yet powerful visualization tools.
We can use `plotCellVsReference()` to manually check how an individual cell in our test set compares to a reference cell of the same or different type. 
This function creates a scatter plot of the chosen cell's expression profile against the specified reference where each point is an individual gene, along with the linear regression fit and Spearman's rank correlation coefficient.

```{r}
# Get first cell label from our predictions - "acinar" in this case.
pred$labels[1]

# Get all cells in reference set with "acinar" label and plot against the first
# in this list.
same.type <- grep(pred$labels[1], sceM$label)
plotCellVsReference(sceG, test.id = 1, ref = sceM, ref.id = same.type[1])

# And now compare to a cell of a different type than our cell of interest.
diff.type <- seq_along(pred$labels)[-same.type]
plotCellVsReference(sceG, test.id = 1, ref = sceM, ref.id = diff.type[1])
```

Another useful function is `plotScoreHeatmap()`, which visualizes the scores for all cells across all reference labels.
This allows users to inspect the confidence of the predicted labels across the dataset.
We can also display clusters (or other metadata information) for each cell by setting `clusters=` or `annotation_col=`.
In this case, we display which donor the cells came from.

```{r}
plotScoreHeatmap(pred, annotation_col=data.frame(donor=sceG$donor,
    row.names=rownames(pred)))
```

## Pruning of scores

The `pruneScores()` function will remove potentially poor-quality or ambiguous assignments.
In particular, ambiguous assignments are identified from cells where the top score is not appreciably better than the median (before fine-tuning) or the next-best score (after fine-tuning).
We also remove cells that have lower top scores than other cells that have been assigned the same label, which may represent a misassignment from a different cell subpopulation.

```{r}
to.remove <- pruneScores(pred)
summary(to.remove)
```

By default, `SingleR()` will automatically report pruned labels where low-quality assignments are replaced with `NA`.
However, the default pruning pruning may not be appropriate for every dataset - see `?pruneScores` for a more detailed discussion.
We provide a few visualization functions to help in determining whether cutoffs might need adjustment.

The first cut-off is a per-cell check that the maximum score associated with a cell (prior to fine-tuning) is greater than the median score for that cell by `min.diff.med`.
We can use `plotScoresSingleCell()` to determine whether this parameter is set appropriately.
Here, we can see that the maximal score was less than 0.05 greater than the median score of that cell.

```{r}
# For a non-pruned label:
plotScoresSingleCell(pred, cell.id = 1)

# For a pruned label:
pruned.cell.ids <- seq_len(nrow(pred))[pruneScores(pred)]
plotScoresSingleCell(pred, cell.id = pruned.cell.ids[1])
```

The second cutoff is a per-label check where low outliers within each label are removed, based on whether they are more than `nmads` median absolute deviations (MADs) below the median score for that label.
Again, several functions can help determine if this parameter needs to be adjusted:

```{r}
plotScoresMultiLabels(pred, dots.on.top = TRUE)
plotScoresSingleLabel(pred, label = "beta", dots.on.top = TRUE)
```

If some tuning parameters must be adjusted, we can simply call `pruneScores()` directly with adjusted parameters.
Here, we set labels to `NA` if they are to be discarded, which is also how `SingleR()` marks such labels in `pruned.labels`.

```{r}
new.pruned <- pred$labels
new.pruned[pruneScores(pred, nmads=5, min.diff.med=0.01)] <- NA
table(new.pruned, useNA="always")
```

# Available reference data sets

## Overview

The [legacy SingleR package](https://github.com/dviraran/SingleR/tree/master/data) provides RDA files that contain normalized expression values and cell types labels based on
bulk RNA-seq, microarray and single-cell RNA-seq data from:

* Blueprint and Encode (`blueprint_encode`),
* the Human Primary Cell Atlas (`hpca`),
* the murine [ImmGen](http://www.immgen.org/) (`immgen`), and
* a collection of mouse data sets downloaded from GEO (`mouse.rna.seq`).

The bulk RNA-seq and microarray data sets of the first three reference data sets were obtained from pre-sorted cell populations, i.e. the cell labels of these samples were mostly derived based on the respective sorting/purification strategy, not via *in silico* prediction methods.

Three additional reference datasets from bulk RNA-seq and microarray data for immune cells have also been prepared. 
Each of these datasets were also obtained from pre-sorted cell populations:

* The [Database for Immune Cell Expression(/eQTLs/Epigenomics)](https://dice-database.org) [@diceRef],
* Novershtern Hematopoietic Cell Data - [GSE24759](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE24759) - formerly known as Differentiation Map [@dmapRef], and
* Monaco Immune Cell Data - [GSE107011](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107011)  [@monaco_immuneRef].

| Data retrieval |  Organism  | Samples | Sample types |  No. of main labels  | No. of fine labels | Cell type focus |
|------------------|----------|----------|-------------|----------------------|------------|----------|
|`HumanPrimaryCellAtlasData()`| human | 713 | microarrays of sorted cell populations  | 37 |  157 | Non-specific |
|`BlueprintEncodeData()` |  human | 259 | RNA-seq | 24 | 43 | Non-specific |
|`DatabaseImmuneCellExpressionData()` | human | 1561 | RNA-seq | 5 | 15 | Immune |
|`NovershternHematopoieticData()` | human | 211 | microarrays of sorted cell populations | 17 | 38 | Hematopoietic & Immune |
|`MonacoImmuneData()` | human | 114 | RNA-seq | 11 | 29 | Immune |
|`ImmGenData()`|  mouse | 830  | microarrays of sorted cell populations | 20 | 253 | Hematopoietic & Immune |
|`MouseRNAseqData()`| mouse |358  |RNA-seq| 18  | 28 | Non-specific |


## Mouse reference datasets

### The Immunological Genome Project (ImmGen)

> The Immunological Genome Project is a collaborative group of immunologists and computational biologists who are generating a complete microarray dissection of gene expression and its regulation in the immune system of the mouse. The project encompasses the innate and adaptive immune systems, surveying all cell types of the myeloid and lymphoid lineages with a focus on primary cells directly *ex vivo*. [Source](http://www.immgen.org/)

**830 microarray samples** of ImmGen raw expression data (phases 1 and 2) were downloaded as CEL files from the Gene Expression Omnibus (GEO; GSE15907 and GSE37448), processed, and normalized 
using the robust multi-array average (RMA) procedure on probe-level data using Matlab functions.
The analysis was performed using a custom CDF file obtained from Brainarray  [@aran2019reference].

There are **20 main cell types** and **253 "fine" cell subtypes**.

<details>
  <summary>Click for Dataset Cell Types</summary>
  
| Label type | Labels |
|------------|--------|
| main | B cells, B cells, pro, Basophils, DC, Endothelial cells, Eosinophils, Epithelial cells, Fibroblasts, ILC, Macrophages, Mast cells, Microglia, Monocytes, Neutrophils, NK cells, NKT, Stem cells, Stromal cells, T cells, Tgd |
| fine | e.g. DC (DC.11B-), DC (DC.103-11B+24+), Macrophages (MF.103-11B+.SALM3, Macrophages (MF.11CLOSER.SALM3), Tgd (Tgd.vg2+.act)|
</details>

### Mouse RNA-seq samples of pure cell populations

The dataset entails **358 mouse RNA-seq** samples annotated to **18 main cell types** and
**28 "fine" cell subtypes**.

<details>
  <summary>Click for Dataset Cell Types</summary>

* Adipocytes
  - Adipocytes
* Astrocytes
  - aNSCs 
  - Astrocytes 
  - Astrocytes activated
* B cells
  - B cells
* Cardiomyocytes
  - Cardiomyocytes
* Dendritic cells
  - Dendritic cells
* Endothelial cells
  - Endothelial cells
* Epithelial cells
  - Ependymal
* Erythrocytes
  - Erythrocytes
* Fibroblasts
  - Fibroblasts
  - Fibroblasts activated
  - Fibroblasts senescent
* Granulocytes
  - Granulocytes
* Hepatocytes
  - Hepatocytes
* Macrophages
  - Macrophages
  - Macrophages activated
* Microglia
  - Microglia
  - Microglia activated
* Monocytes
  - Monocytes
* Neurons
  - Neurons
  - Neurons activated
  - NPCs
  - qNSCs
* NK cells 
  - NK cells
* Oligodendrocytes
  - Oligodendrocytes
  - OPCs
* T cells
  - T cells

</details>

The RNA-seq data were downloaded from GEO and processed by the Benayoun Lab as follows [@Benayoun2019]:

>Paired-end 75-bp reads were trimmed using Trim Galore! 0.3.1 (github.com/FelixKrueger/TrimGalore) to retain high-quality bases with phred score >15 and a remaining length >35 bp. Read pairs were mapped to the UCSC mm9 genome build using STAR 2.4.0j (Dobin et al. 2013). Read counts were assigned to genes using subread 1.4.5-p1 (Liao et al. 2014) and were imported into R to estimate differential gene expression as a function of age using the DESeq2 R package (DESeq2 1.6.3). 

More details can be found in the [Benayoun Lab github repo](https://github.com/BenayounLaboratory/Mouse_Aging_Epigenomics_2018/tree/master/FigureS7_CIBERSORT).

## Human reference datasets

### Blueprint and ENCODE

Blueprint: 144 bulk RNA-seq of pure immune samples annotated to 28 cell types [@blueprintRef]

ENCODE: 115 RNA-seq pure stroma and immune samples annotated to 17 cell types [@encodeRef]

Altogether, **259 samples** with **24 main cell types** and  **43 "fine" cell subtypes**.

<details>
  <summary>Click for Dataset Cell Types</summary>
  

* Adipocytes
  - Adipocytes
  - Astrocytes
  - Preadipocytes
* B-cells
  - Class-switched memory B-cells
  - Memory B-cells
  - naive B-cells
  - Plasma cells
* CD4+ T-cells
  - CD4+ T-cells 
  - CD4+ Tcm 
  - CD4+ Tem
  - Tregs
* CD8+ T-cells
  - CD8+ T-cells 
  - CD8+ Tcm 
  - CD8+ Tem
* Chondrocytes
  - Chondrocytes
* DC 
  - DC
* Endothelial cells
  - Endothelial cells
  - mv Endothelial cells
* Eosinophils
  - Eosinophils
* Epithelial cells 
  - Epithelial cells
* Erythrocytes
  - Erythrocytes
* Fibroblasts
  - Fibroblasts
* HSC
  - CLP 
  - CMP
  - GMP 
  - HSC
  - Megakaryocytes
  - MEP
  - MPP
* Keratinocytes
  - Keratinocytes
* Macrophages
  - Macrophages 
  - Macrophages M1 
  - Macrophages M2
* Melanocytes
  - Melanocytes
* Mesangial cells
  - Mesangial cells
* Monocytes
  - Monocytes
* Myocytes
  - Myocytes
* Neurons
  - Neurons
* Neutrophils
  - Neutrophils
* NK cells
  - NK cells
* Pericytes
  - Pericytes
* Skeletal muscle
  - Skeletal muscle
* Smooth muscle 
  - Smooth muscle

</details>

### Human Primary Cell Atlas

Human Primary Cell Atlas (HPCA): a collection of Gene Expression Omnibus (GEO datasets), which
contains **713 microarray samples** classified to **37 main cell types** and further annotated to **157  "fine" cell subtypes** [@hpcaRef].

<details>
  <summary>Click for Dataset Cell Types</summary>
  
| Label type | Labels |
|------------|--------|
| main | Astrocyte, B_cell, BM, BM & Prog., Chondrocytes, CMP, DC, Embryonic_stem_cells, Endothelial_cells, Epithelial_cells, Erythroblast, Fibroblasts, Gametocytes, GMP, Hepatocytes, HSC_-G-CSF, HSC_CD34+, iPS_cells, Keratinocytes, Macrophage, MEP, Monocyte, MSC, Myelocyte, Neuroepithelial_cell, Neurons, Neutrophil, Neutrophils, NK_cell, Osteoblasts, Platelets, Pre-B_cell_CD34-, Pro-B_cell_CD34+, Pro-Myelocyte, Smooth_muscle_cells, T_cells, Tissue_stem_cells |
| fine | e.g. DC:monocyte-derived:AM580, DC:monocyte-derived:Poly(IC), T_cell:CD8+_naive, Keratinocytes:IL22, Keratinocytes:KGF |
</details>

### Database of Immune Cell Expression

[DICE](https://dice-database.org): **1561 bulk RNA-seq samples** of 15 immune cell types [@diceRef].
Samples were annotated to **5 main cell types** and **15 "fine" cell subtypes**.


<details>
  <summary>Click for Dataset Cell Types</summary>
  
* B cells
  - B cells, naive
* Monocytes
  - Monocytes, CD14+
  - Monocytes, CD16+
* NK cells
  - NK cells
* T cells, CD8+
  - T cells, CD8+, naive 
  - T cells, CD8+, naive, stimulated
* T cells, CD4+
  - T cells, memory TREG 
  - T cells, CD4+, naive 
  - T cells, CD4+, naive, stimulated 
  - T cells, CD4+, naive Treg 
  - T cells, CD4+, Th1 
  - T cells, CD4+, Th1_Th17 
  - T cells, CD4+, Th2  
  - T cells, CD4+, TFH 
  - T cells, CD4+, Th17
  
</details> 

### Novershtern Hematopoietic Data (Differentiation Map)

This dataset is composed of **211 microarray samples** from hematopoietic cell subsets as described in [@dmapRef].
Samples were annotated to **16 main cell types** and **38 "fine" cell subtypes**. 

<details>
  <summary>Click for Dataset Cell Types</summary>
  
* Basophils
  - Basophils
* B cells
  - Naive B cells
  - Mature B-cells class able to switch
  - Mature B-cells
  - Mature B-cells class switched
  - Early B-cells
  - Pro B-cells
* CMPs
  - Common myeloid progenitors
* Dendritic cells
  - Plasmacytoid Dendritic Cells
  - Myeloid Dendritic Cells
* Eosinophils
  - Eosinophils
* Erythroid cells
  - Erythroid_CD34+ CD71+ GlyA-
  - Erythroid_CD34- CD71+ GlyA-
  - Erythroid_CD34- CD71+ GlyA+
  - Erythroid_CD34- CD71lo GlyA+
  - Erythroid_CD34- CD71- GlyA+
* GMPs
  - Granulocyte/monocyte progenitors
* Granulocytes
  - Colony Forming Unit-Granulocytes
  - Granulocyte (Neutrophilic Metamyelocytes)
  - Granulocyte (Neutrophils)
* HSCs
  - Hematopoietic stem cells_CD133+ CD34dim
  - Hematopoietic stem cell_CD38- CD34+
* Megakaryocytes
  - Colony Forming Unit-Megakaryocytic
  - Megakaryocytes
* MEPs
  - Megakaryocyte/erythroid progenitors
* Monocytes
  - Colony Forming Unit-Monocytes, Monocytes
  - Monocytes
* NK cells
  - Mature NK cells_CD56- CD16+ CD3-
  - Mature NK cells_CD56+ CD16+ CD3-
  - Mature NK cells_CD56- CD16- CD3-
* NK T cells
  - NK T cells
* CD8+ T cells
  - CD8+ Effector Memory RA
  - Naive CD8+ T-cells
  - CD8+ Effector Memory
  - CD8+ Central Memory
* CD4+ T cells
  - Naive CD4+ T-cells
  - CD4+ Effector Memory
  - CD4+ Central Memory

</details>

### Monaco Immune Data

This dataset is composed of **114 bulk RNA-seq samples** immune cell subsets as described in [@monaco_immuneRef].
Samples were annotated to **11 main cell types** and **29 "fine" cell subtypes**. 

<details>
  <summary>Click for Dataset Details</summary>
  
* CD8+ T cells
  - Tcell.CD8.naive 
  - Tcell.CD8.CentralMemory 
  - Tcell.CD8.EffectorMemory 
  - Tcell.CD8.TerminalEffector 
* T cells
  - Tcell.MAIT 
  - Tcell.VD2pos.gd 
  - Tcell.VD2neg.gd 
* CD4+ T cells
  - Tcell.FollicularHelper 
  - Tcell.Treg 
  - Tcell.Th1 
  - Tcell.Th1.Th17 
  - Tcell.Th17 
  - Tcell.Th2 
  - Tcell.CD4.naive 
  - Tcell.CD4.TerminalEffector
* Progenitors
  - Progenitor
* B cells
  - Bcell.naive 
  - Bcell.NonSwitchedMemory 
  - Bcell.Exhausted 
  - Bcell.SwitchedMemory
* Plasmablasts
  - Plasmablast
* Monocytes
  - Monocyte.Classical 
  - Monocyte.Int 
  - Monocyte.NonClassical
* NK cells
  - NK
* Dendritic cells
  - Dendritic.Plasmacytoid 
  - Dendritic.Myeloid
* Neutrophils
  - Neutrophil
* Basophils 
  - Basophil

</details>

# Separating training from classification

Advanced users can split the `SingleR()` workflow into two separate training and classification steps.
This means that training (e.g., marker detection, assembling of nearest-neighbor indices) only needs to be performed once.
The resulting data structures can then be re-used across multiple classifications with different test datasets, provided the test feature set is identical to or a superset of the features in the training set.
For example:

```{r}
trained <- trainSingleR(sceM, labels=sceM$label, genes=markers)
pred2b <- classifySingleR(sceG, trained)
table(pred$labels, pred2b$labels)
```

Other efficiency improvements are possible through several arguments:

- Switching to an approximate algorithm for the nearest neighbor search in `trainSingleR()` via the `BNPARAM=` argument from the `r Biocpkg("BiocNeighbors")` package.
- Parallelizing the fine-tuning step in `classifySingleR()` with the `BPPARAM=` argument from the `r Biocpkg("BiocParallel")` package.

These arguments can also be specified in the `SingleR()` command.

# Session information

```{r}
sessionInfo()
```

# References
