---
title: "Obtaining reference data from GSE107011"
author: 
- name: Jared M. Andrews
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
date: "Revised: September 5th, 2019"
output: 
  BiocStyle::html_document
---

```{r setup, echo=FALSE, include=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Overview

This RNA-seq dataset was downloaded from [GSE107011](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107011).
Every sample represents the transcriptome of a specific cell type; this data is therefore well suited to be used as a general training data set for the typical _SingleR_ analysis.
Values were already TPM normalized, so the only additional processing done was to remove 'PBMC' samples, remove genes with no reads across samples, collapse duplicate genes, and log2 normalize values.
'main' and 'fine' labels were manually assigned to each sample based on cell type as specified in the GEO repository.

# Data retrieval

First, we'll download the TPM normalized values from GEO.

```{r retrieve_from_geo}
library(BiocFileCache)
url <- "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE107nnn/GSE107011/suppl/GSE107011_Processed_data_TPM.txt.gz"
bfc <- BiocFileCache(ask=FALSE)
ref <- bfcrpath(bfc, url)
GSE107011 <- as.matrix(read.table(ref, sep = "\t", header = TRUE, row.names = 1))
```

# Data processing
We don't want the `PBMC` samples, as they aren't purified cell types and don't make for good reference data.

```{r drop_pbmc}
pbmc <- grep("PBMC", colnames(GSE107011))
GSE107011 <- GSE107011[, -pbmc]
```

Genes with no reads in any samples don't provide any value for our purpose, so we will remove those as well.

```{r drop_no_read_rows}
library(matrixStats)
GSE107011 <- GSE107011[rowSums(GSE107011) != 0, ]
```

This data has Ensembl gene IDs, but we need gene symbols, which can be acquired via `r Biocpkg("biomaRt")`.

```{r get_symbols}
library("biomaRt")
genes <- rownames(GSE107011)
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")
res <- getBM(attributes = c("ensembl_gene_id_version", "hgnc_symbol"), 
             filters = "ensembl_gene_id_version", values = genes, mart = ensembl)
symbols <- res$hgnc_symbol[match(genes, res$ensembl_gene_id_version)]
```

Many Ensembl IDs don't have official gene symbols, so we will remove those as well. 
Then we can assign the gene symbols in place of the Ensembl IDs.

```{r drop_no_symbol_rows}
discard <- symbols == "" | is.na(symbols)
GSE107011 <- GSE107011[!discard,]
rownames(GSE107011) <- symbols[!discard]
```

A small number of genes are duplicated, so those will be collapsed by keeping the instances with the highest median of reads across samples.

```{r collapse_dup_genes}
meds <- rowMedians(GSE107011)
GSE107011 <- GSE107011[order(rownames(GSE107011), -meds),]
GSE107011 <- GSE107011[!duplicated(rownames(GSE107011)), ]
```

Then we will log2 normalize after adding a pseudocount of 1.

```{r log_normalize}
GSE107011 <- GSE107011 + 1
logcounts <- log2(GSE107011)
```

# Sample labelling
We can now apply human-readable labels to each sample.
This requires some translation.

```{r create_metadata}
library(GEOquery)
GSE107011 <- getGEO("GSE107011")
fine <- pData(GSE107011[[1]])[["cell type:ch1"]]
fine <- fine[-pbmc]
fine <- sub("cell$", "cells", fine)

dictionary <- c(
	`Naive CD8 T cells`="CD8+ T cells",
	`Central memory CD8 T cells`="CD8+ T cells",
	`Effector memory CD8 T cells`="CD8+ T cells",
	`Terminal effector CD8 T cells`="CD8+ T cells",
	`MAIT cells`="T cells",
	`Vd2 gd T cells`="T cells",
	`Non-Vd2 gd T cells`="T cells",
	`Follicular helper T cells`="CD4+ T cells",
	`T regulatory cells`="CD4+ T cells",
	`Th1 cells`="CD4+ T cells",
	`Th1/Th17 cells`="CD4+ T cells",
	`Th17 cells`="CD4+ T cells",
	`Th2 cells`="CD4+ T cells",
	`Naive CD4 T cells`="CD4+ T cells",
	`Progenitor cells`="Progenitors",
	`Naive B cells`="B cells",
	`Non-switched memory B cells`="B cells",
	`Exhausted B cells`="B cells",
	`Switched memory B cells`="B cells",
	`Plasmablasts`="B cells",
	`Classical monocytes`="Monocytes",
	`Intermediate monocytes`="Monocytes",
	`Non classical monocytes`="Monocytes",
	`Natural killer cells`="NK cells",
	`Plasmacytoid dendritic cells`="Dendritic cells",
	`Myeloid dendritic cells`="Dendritic cells",
	`Low-density neutrophils`="Neutrophils",
	`Low-density basophils`="Basophils",
	`Terminal effector CD4 T cells`="CD4+ T cells"
)

main <- dictionary[fine]
stopifnot(all(!is.na(main)))

library(S4Vectors)
coldata <- DataFrame(row.names = colnames(logcounts),
                     label.main = main,
                     label.fine = fine)
```


# Saving to file

Now the counts and metadata can be saved for upload to `r Biocpkg("ExperimentHub")`.

```{r save_for_ExpHub}
path <- file.path("SingleR", "monaco_immune", "1.0.0")
dir.create(path, showWarnings = FALSE, recursive = TRUE)

saveRDS(logcounts, file = file.path(path, "logcounts.rds"))
saveRDS(coldata, file = file.path(path, "coldata.rds"))
```

# Session info

```{r}
sessionInfo()
```
