df <- data.frame(
    Title = sprintf("DMAP RNA microarray %s", c("logcounts", "colData")), # This can be the exact file name (if self-describing) or a more complete description.
    Description = sprintf("%s 211 microarray samples of hematopoietic cells generated by Differentiation Map (GSE24759)", 
        c("Matrix of log-normalized expression values from", "Per-sample metadata containing cell type labels of")),
    RDataPath = file.path("SingleR", "dmap","1.0.0", c("logcounts.rds", "coldata.rds")),
    BiocVersion="3.10", # The first Bioconductor version the resource was made available for.
    Genome=NA, # Can be NA.
    SourceType="TXT", #  Format of original data, e.g., FASTA, BAM, BigWig, etc. ‘getValidSourceTypes()’ for currently acceptable values
    SourceUrl="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE24759",
    SourceVersion="1.0.0",
    Species="Homo sapiens", # getSpeciesList, validSpecies, or suggestSpecies(); can be NA
    TaxonomyId="9606",
    Coordinate_1_based=NA, #TRUE, FALSE, NA
    DataProvider="GEO",
    Maintainer="Jared Andrews <jared.andrews@wustl.edu>",
    RDataClass="matrix", # R / Bioconductor class the data are stored in, e.g., GRanges, SummarizedExperiment, ExpressionSet etc.
    DispatchClass="Rds", # Determines how data are loaded into R.
    stringsAsFactors = FALSE
)

write.csv(file="../extdata/metadata-dmap.csv", df, row.names=FALSE)
