########################################################################
### Functions for the retrieval of individual data sets from ExpHub
########################################################################

#' Obtain the HPCA data
#'
#' Download and cache the normalized expression values of the data stored in
#' the Human Primary Cell Atlas. The data will be downloaded from ExperimentHub,
#' returning a \linkS4class{SummarizedExperiment} object for further use.
#'
#' @details
#' This function provides normalized expression values 713 microarray samples of 
#' the Human Primary Cell Atlas (HPCA) (Mabbott et al., 2013).
#' These 713 samples were processed and normalized as described in Aran, Looney and 
#' Liu et al. (2019) and each sample has been assigned to one of 37 main cell types
#' and 157 subtypes.
#'
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Friederike Dündar
#'
#' @references
#' Mabbott NA et al. (2013).
#' An expression atlas of human primary cells: inference of gene function from coexpression networks.
#' \emph{BMC Genomics} 14, Article 632. 
#' 
#' Aran D, Looney AP, Liu L et al. (2019). 
#' Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.
#' \emph{Nat. Immunol.} 20, 163–172. 
#' 
#' @examples
#' ref.se <- HumanPrimaryCellAtlasData()
#'
#' @export
#' @importFrom SummarizedExperiment rowData
HumanPrimaryCellAtlasData <- function() {
    version <- "1.0.0"
    .create_se(file.path("hpca", version),
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)
}

#' Obtain human bulk RNA-seq data from Blueprint and ENCODE
#'
#' Download and cache the normalized expression values of 259 RNA-seq samples of
#' pure stroma and immune cells as generated and supplied by Blueprint and ENCODE.
#'
#' @details
#' This function provides normalized expression values of 259 bulk RNA-seq samples
#' generated by Blueprint and ENCODE from pure populations of stroma and immune 
#' cells (Martens and Stunnenberg, 2013; The ENCODE Consortium, 2012).
#'
#'  Blueprint Epigenomics: 144 RNA-seq pure immune samples annotated to 28 cell types.
#'  ENCODE: 115 RNA-seq pure stroma and immune samples annotated to 17 cell types.
#'  Altogether, 259 samples with 43 cell types.
#' 
#' The samples were processed and normalized as described in Aran, Looney and
#' Liu et al. (2019); i.e. the raw RNA-seq counts were downloaded from Blueprint
#' and ENCODE in 2016 and normalized via edgeR (TPMs).
#' 
#' The data will be downloaded from ExperimentHub,
#' returning a \linkS4class{SummarizedExperiment} object for further use.
#'
#' @param rm.NA String specifying how missing values should be handled.
#' \code{"rows"} will remove genes with at least one missing value, 
#' \code{"cols"} will remove samples with at least one missing value,
#' \code{"both"} will remove any gene or sample with at least one missing value,
#' and \code{"none"} will not perform any removal.
#'
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Friederike Dündar
#'
#' @references
#' The ENCODE Project Consortium (2012).
#' An integrated encyclopedia of DNA elements in the human genome.
#' \emph{Nature} 489, pages 57–74.
#' 
#' Martens JHA and Stunnenberg HG (2013). 
#' BLUEPRINT: mapping human blood cell epigenomes.
#' \emph{Haematologica} 98, 1487–1489.
#' 
#' Aran D, Looney AP, Liu L et al. (2019). 
#' Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.
#' \emph{Nat. Immunol.} 20, 163–172. 
#' 
#' @examples
#' ref.se <- BlueprintEncodeData(rm.NA = "rows")
#'
#' @export
BlueprintEncodeData <- function(rm.NA = c("rows","cols","both","none")) {
    version <- "1.0.0"
    rm.NA <- match.arg(rm.NA)
    .create_se(file.path("blueprint_encode", version), 
        assays="logcounts", rm.NA = rm.NA,
        has.rowdata = FALSE, has.coldata = TRUE)
}

#' Obtain mouse bulk expression data from the Immunologic Genome Project
#'
#' Download and cache the normalized expression values of 830 microarray samples of
#' pure mouse immune cells, generated by the Immunologic Genome Project (ImmGen).
#'
#' @details
#' This function provides normalized expression values of 830 microarray samples
#' generated by ImmGen from pure populations of murine immune cells (<http://www.immgen.org/>).
#' 
#' The samples were processed and normalized as described in Aran, Looney and
#' Liu et al. (2019); i.e. CEL files from the Gene Expression Omnibus (GEO; GSE15907 and GSE37448), 
#' were downloaded, processed, and normalized using the robust multi-array average
#' (RMA) procedure on probe-level data (with Matlab functions).
#' 
#' The data will be downloaded from ExperimentHub,
#' returning a \linkS4class{SummarizedExperiment} object for further use.
#'
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Friederike Dündar
#' 
#' @references
#' Heng TS, Painter MW, Immunological Genome Project Consortium (2008).
#' The Immunological Genome Project: networks of gene expression in immune cells.
#' \emph{Nat. Immunol.} 9, 1091-1094. 
#' 
#' Aran D, Looney AP, Liu L et al. (2019). 
#' Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.
#' \emph{Nat. Immunol.} 20, 163–172. 
#' 
#' @examples
#' ref.se <- ImmGenData()
#' 
#' @export
ImmGenData <- function() {
    version <- "1.0.0"
    .create_se(file.path("immgen", version), 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)
}

#' Obtain mouse bulk expression data of sorted cell populations (RNA-seq)
#'
#' Download and cache the normalized expression values of 358 bulk RNA-seq samples
#' of sorted cell populations that can be found at GEO.
#'
#' @details This dataset was contributed by the Benayoun Lab that identified, 
#' downloaded and processed data sets on GEO that corresponded to sorted cell
#' types (Benayoun et al., 2019).
#' 
#' The dataset entails 358 mouse RNA-seq samples annotated to 18 main cell types: 
#' Adipocytes, Astrocytes, B cells, Cardiomyocytes, Dendritic cells, Endothelial 
#' cells, Epithelial cells, Erythrocytes, Fibroblasts, Granulocytes, Hepatocytes,
#' Macrophages, Microglia, Monocytes, Neurons, NK cells, Oligodendrocytes, T cells.
#' 
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Friederike Dündar
#' 
#' @references
#' Benayoun B et al. (2019).
#' Remodeling of epigenome and transcriptome landscapes with aging in mice reveals widespread induction of inflammatory responses.
#' \emph{Genome Res.} 29, 697-709.
#' 
#' Code at \url{https://github.com/BenayounLaboratory/Mouse_Aging_Epigenomics_2018/tree/master/FigureS7_CIBERSORT/RNAseq_datasets_for_Deconvolution/2017-01-18}
#' 
#' @examples
#' ref.se <- MouseRNAseqData()
#' 
#' @export
MouseRNAseqData <- function() {
    version <- "1.0.0"
    .create_se(file.path("mouse.rnaseq", version), 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)
}

#' Obtain human bulk RNA-seq from the Database of Immune Cell Expression (DICE)
#'
#' Download and cache the normalized expression values of 1561 bulk RNA-seq samples
#' of sorted cell populations from the DICE database.
#'
#' @details 
#' This function provides normalized expression values of 1561 bulk RNA-seq samples
#' generated by DICE from pure populations of human immune cells (<http://www.dice-database.org/>).
#'
#' TPM normalized values for each cell type were downloaded from \url{https://dice-database.org/downloads}. 
#' Genes with no reads across samples were removed, and values were log2 normalized after a pseudocount of 1 was added.
#' 
#' The dataset entails 1561 human RNA-seq samples annotated to 5 main cell types: 
#' B cells, Monocytes, NK cells, CD8+ T cells, and CD4+ T cells.
#'
#' It includes 15 total cell types:
#' 
#' 
#' 
#' 
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Jared Andrews
#' 
#' @references
#' Schmiedel B et al. (2018).
#' Impact of Genetic Polymorphisms on Human Immune Cell Gene Expression.
#' \emph{Cell} 175, 1701-1715.
#' 
#' @examples
#' ref.se <- DatabaseImmuneCellExpressionData()
#' 
#' @export
DatabaseImmuneCellExpressionData <- function() {
    version <- "1.0.0"
    .create_se(file.path("dice", version), 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)
}

#' Obtain human bulk expression data of sorted cell populations from GSE24759 (microarray)
#'
#' Download and cache the normalized expression values of 211 bulk human microarray samples
#' of sorted cell populations that can be found in GEO.
#'
#' @details This dataset was contributed by the Benayoun Lab that identified, 
#' downloaded and processed data sets on GEO that corresponded to sorted cell
#' types (Benayoun et al., 2019).
#' 
#' The dataset entails 211 human microarray samples annotated to 18 main cell types: 
#' Adipocytes, Astrocytes, B cells, Cardiomyocytes, Dendritic cells, Endothelial 
#' cells, Epithelial cells, Erythrocytes, Fibroblasts, Granulocytes, Hepatocytes,
#' Macrophages, Microglia, Monocytes, Neurons, NK cells, Oligodendrocytes, T cells.
#' 
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Jared Andrews
#' 
#' @references
#' Novershtern N et al. (2011).
#' Densely interconnected transcriptional circuits control cell states in human hematopoiesis.
#' \emph{Cell} 144, 296-309.
#' 
#' @examples
#' ref.se <- NovershternHematopoieticData()
#' 
#' @export
NovershternHematopoieticData <- function() {
    version <- "1.0.0"
    .create_se(file.path("dmap", version), 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)
}

#' Obtain human bulk expression data of sorted cell populations from GSE107011 (RNA-seq)
#'
#' Download and cache the normalized expression values of 114 bulk RNA-seq samples
#' of sorted cell populations that can be found in GEO.
#'
#' @details This dataset was contributed by the Benayoun Lab that identified, 
#' downloaded and processed data sets on GEO that corresponded to sorted cell
#' types (Benayoun et al., 2019).
#' 
#' The dataset entails 114 human RNA-seq samples annotated to 11 main cell types: 
#' Adipocytes, Astrocytes, B cells, Cardiomyocytes, Dendritic cells, Endothelial 
#' cells, Epithelial cells, Erythrocytes, Fibroblasts, Granulocytes, Hepatocytes,
#' Macrophages, Microglia, Monocytes, Neurons, NK cells, Oligodendrocytes, T cells.
#' 
#' @return A \linkS4class{SummarizedExperiment} object with a \code{"logcounts"} assay
#' containing the log-normalized expression values, along with cell type labels in the 
#' \code{\link{colData}}.
#'
#' @author Jared Andrews
#' 
#' @references
#' Monaco G et al. (2019).
#' RNA-Seq Signatures Normalized by mRNA Abundance Allow Absolute Deconvolution of Human Immune Cell Types
#' \emph{Cell Rep.} 26, 1627-1640.
#' 
#' @examples
#' ref.se <- MonacoImmuneData()
#' 
#' @export
MonacoImmuneData <- function() {
    version <- "1.0.0"
    .create_se(file.path("monaco_immune", version), 
        assays="logcounts", rm.NA = "none",
        has.rowdata = FALSE, has.coldata = TRUE)
}

#####################################################################
### Helper function
#####################################################################

#' @importFrom ExperimentHub ExperimentHub
#' @importFrom SummarizedExperiment SummarizedExperiment
#' @importFrom SummarizedExperiment rowData
#' @importFrom BiocFileCache BiocFileCache bfcrpath
#' @importFrom S4Vectors DataFrame
.create_se <- function(dataset, hub = ExperimentHub(), assays="logcounts",
    rm.NA = c("rows","cols","both","none"), has.rowdata=FALSE, has.coldata=TRUE) 
{
    rm.NA <- match.arg(rm.NA)
    
    ## TEMPORARY CODE for pulling data directly from github until it is 
    ## available on ExHub =====================================================
    if (dirname(dataset) == "dice") {
        full.url <- "https://www.dropbox.com/s/qzh0k3g0ulihpyd/dice.rda?dl=1"
    } else if (dirname(dataset) == "dmap") {
        full.url <- "https://www.dropbox.com/s/fz118owqdm6bty2/dmap.rda?dl=1"
    } else if (dirname(dataset) == "monaco_immune") {
        full.url <- "https://www.dropbox.com/s/3l82mnuvp3ts4gt/monaco_immune.rda?dl=1"
    } else {
        full.url <- sprintf("https://github.com/dviraran/SingleR/blob/master/data/%s.rda?raw=true", dirname(dataset))
    }

    bfc <- BiocFileCache(ask=FALSE)
    ref <- bfcrpath(bfc, full.url)

    env <- new.env()
    load(ref, envir = env)
    ref.set <- get(dirname(dataset), envir = env)

    nrmcnts <- ref.set$data
    nrmcnts <- .rm_NAs(nrmcnts, rm.NA)

    new.names <- colnames(nrmcnts)
    indices <- split(seq_along(new.names), new.names)
    for (j in names(indices)) {
        idx <- indices[[j]]
        if (length(idx) > 1L) {
            new.names[idx] <- sprintf("%s.%i", j, seq_along(idx))
        }
    }
    colnames(nrmcnts) <- new.names

    coldata <- DataFrame(row.names = colnames(nrmcnts),
        label.main = ref.set$main_types,
        label.fine = ref.set$types)
    ref.se <- SummarizedExperiment(assays = list(logcounts = nrmcnts),
        colData = coldata)
    
    return(ref.se)
    
    ##########################################################################
    ## code for when data is actually available on ExpHub ===================
    ##########################################################################
    host <- file.path("SingleR", dataset)
    
    ## extract normalized values --------
    all.assays <- list()
    for (a in assays) {
        nrmcnts <- hub[hub$rdatapath==file.path(host, paste0(a, ".rds"))][[1]]
        all.assays[[a]] <- .rm_NAs(nrmcnts, rm.NA)
    }
    
    ## get metadata ----------------------
    args <- list()
    if (has.coldata) {
        args$colData <- hub[hub$rdatapath==file.path(host, "coldata.rds")][[1]]
    }
    if (has.rowdata) {
        args$rowData <- hub[hub$rdatapath==file.path(host, "rowdata.rds")][[1]]
    }
    
    ## make the final SE object ----------
    do.call(SummarizedExperiment, c(list(assays=all.assays), args))
}

#' @importFrom DelayedMatrixStats rowAnyNAs colAnyNAs
#' @importFrom DelayedArray DelayedArray
.rm_NAs <- function(mat, rm.NA = "rows"){
    # Identify them first before removal, to ensure that 
    # the same rows are columns are removed with 'both'.
    if(rm.NA == "rows" || rm.NA == "both"){
        keep_rows <- !rowAnyNAs(DelayedArray(mat))
    } else {
        keep_rows <- !logical(nrow(mat))
    }

    if (rm.NA=="cols" || rm.NA == "both") {
        keep_cols <- !colAnyNAs(DelayedArray(mat))
    } else {
        keep_cols <- !logical(ncol(mat))
    }

    # Avoid making unnecessary copies if possible.
    if (!all(keep_rows)) {
        mat <- mat[keep_rows,,drop=FALSE]
    }
    if (!all(keep_cols)) {
        mat <- mat[,keep_cols,drop=FALSE]
    }
    mat
}
